posts/bonus-schema-guru.md

```diff
@@ -0,0 +1,233 @@
+
+### Validate and serialize JSON data
+
+The last new feature added to Fastify is the ability to validate and serialize JSON data. This one is just a shortcut to write even less code:
+
+```js
+fastify.post('/light-foo',
+  async function handler (request, reply) {
+    const validationResult = request.validate({ name: 'John' }, bodySchema)
+    console.log(validationResult)
+
+    const serializedResponse = reply.serializeInput({ user: { id: 1 } }, responseSchema)
+    return serializedResponse
+  })
+```
+
+## Performance
+
+Fastify is always thinking about the performance of your application. The new features we just read above, have this mindset too!
+
+- `getValidationFunction`
+- `compileValidationSchema`
+- `validate`
+- `getSerializationFunction`
+- `compileSerializationSchema`
+- `serializeInput`
+
+All the new functions implements a `WeakMap` cache to avoid to recompile the functions every time.
+Compiling a new function is a very expensive operation, so it may impact your application performance.
+
+To avoid this, Fastify caches the compiled functions in a `WeakMap` so, every time you call the above functions with the same JSON schema, it will use the cached version.
+
+There is a foundamental thing you must do to benefit from this cache: reuse the same JSON schema objects.
+Let's see how to do it!
+
+### The WeakMap cache
+
+To understand how the cache works, let's take a look at the code:
+
+```js
+const cache = new WeakMap()
+
+const schema = { id: 'foo' }
+cache.set(schema, function () { console.log('foo') })
+
+console.log(cache.has({ id: 'foo' })) // prints false
+console.log(cache.has(schema)) // prints true
+```
+
+As you can understand, the `WeakMap` cache is a map that stores the JSON as key and the compiled function as function.
+
+If the key is not the same object reference, the cache will not be able to find the compiled function.
+
+That said, to benefit from the cache, you must use the same JSON schema object:
+
+```js
+const aSchema = {
+  type: 'object',
+  properties: {
+    name: { type: 'string', minLength: 2 }
+  }
+}
+fastify.get('/',
+  async function handler (request, reply) {
+    // we call the compileValidationSchema with the same schema object
+    const validationFn = request.compileValidationSchema(aSchema)
+    const validationResult = validationFn({ name: 'John' })
+    console.log(validationResult)
+    return validationResult
+  })
+```
+
+Instead, by writing a new schema object, the cache will not be able to find the compiled function:
+
+```js
+fastify.get('/',
+  async function handler (request, reply) {
+    // we call the compileValidationSchema with a new JSON object every handler call
+    // so the cache will not be able to find the compiled function
+    const validationFn = request.compileValidationSchema({
+      type: 'object',
+      properties: {
+        name: { type: 'string', minLength: 2 }
+      }
+    })
+    // ...
+  })
+```
+
+
+## Summary
```

Comment:
What do you think about adding also a small paragraph to indicate that if they have any feedback when using it, please provide it in a form of an issue. I'm asking here mostly about the ergonomics, so we can adjust them based on the feedback?